src/config/mongoose_config_parser_toml.erl

```diff
@@ -1522,9 +1597,27 @@ parse_list(Path, L) ->
 -spec handle(path(), toml_value()) -> option().
 handle(Path, Value) ->
     Handler = handler(Path),
-    Option = Handler(Path, Value),
-    mongoose_config_validator_toml:validate(Path, Option),
-    Option.
+    try Handler(Path, Value) of
+        Option ->
+                validate(Path, Option),
+                Option
+    %% attach path if it has no path attached yet
+    catch Class:Error:Stacktrace
+         when not is_map(Error); not is_map_key(path, Error) ->
+              erlang:raise(Class, #{what => toml_parse_failed,
+                                    path => Path,
+                                    reason => Error}, Stacktrace)
+    end.
+
+validate(Path, Option) ->
+    try
+        mongoose_config_validator_toml:validate(Path, Option)
+    catch Class:Error:Stacktrace
+         when not is_map(Error); not is_map_key(path, Error) ->
+              erlang:raise(Class, #{what => toml_validate_failed,
+                                    path => Path,
+                                    reason => Error}, Stacktrace)
+    end.
```

Comment:
IMO this is against the "don't care about errors yet" statement in the JIRA story.

However, I would not vote against a change just because it's out of scope.

There are following problems here:
1. Code complexity - it increase significantly
2. Nesting and repetition - I would rather see both try...catch clauses at least at the same level in code so that one does not have to fall under the other one.
3. Overly generous catching: `_:_:_` is kind of a last resort. Here it should always be the `error` class and we could start from there - not to accidentally catch the `exit` class or the `throw` class.

Anyway, this **will** be rewritten soon, there is a story for user-friendly errors which needs to change this exact piece of code.

Bottom line: at least fix points 2 and 3 - however I do not see the necessity of the whole change here to be honest, even after seeing the examples provided by @NelsonVides  :-)

Reply:
We could reuse code by introducing step function, but it would be in stacktraces:

```erlang
%%% config_parser_SUITE ==> services.service_admin_extra: FAILED
%%% config_parser_SUITE ==> {#{path => [<<"submods">>,<<"service_admin_extra">>,<<"services">>],
   reason => function_clause,what => toml_parse_failed},
 [{lists,flatmap,[#Fun<mongoose_config_parser_toml.22.109831236>,1],[]},
  {mongoose_config_parser_toml,service_opt,2,
                               [{file,"/Users/mikhailuvarov/erlang/esl/MongooseIM/src/config/mongoose_config_parser_toml.erl"},
                                {line,608}]},
  {mongoose_config_parser_toml,try_step,4,
                               [{file,"/Users/mikhailuvarov/erlang/esl/MongooseIM/src/config/mongoose_config_parser_toml.erl"},
                                {line,1607}]},
  {mongoose_config_parser_toml,handle,2,
                               [{file,"/Users/mikhailuvarov/erlang/esl/MongooseIM/src/config/mongoose_config_parser_toml.erl"},
                                {line,1600}]},
  {lists,flatmap,2,[{file,"lists.erl"},{line,1250}]},
  {mongoose_config_parser_toml,process_service,2,
                               [{file,"/Users/mikhailuvarov/erlang/esl/MongooseIM/src/config/mongoose_config_parser_toml.erl"},
                                {line,603}]},
  {mongoose_config_parser_toml,try_step,4,
                               [{file,"/Users/mikhailuvarov/erlang/esl/MongooseIM/src/config/mongoose_config_parser_toml.erl"},
                                {line,1607}]},
  {mongoose_config_parser_toml,handle,2,
                               [{file,"/Users/mikhailuvarov/erlang/esl/MongooseIM/src/config/mongoose_config_parser_toml.erl"},
                                {line,1600}]}]}
```